[cli] Added modules for interacting with Apple and Android platforms

EvanBacon opened the pull request:
# Why

- Split out Platform modules from https://github.com/expo/expo/pull/16160
- Pending https://github.com/expo/expo/pull/16508

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

# How

- Pull in all `cli/start/platforms` code from https://github.com/expo/expo/pull/16160
- Based on the `Simulator` and `Android` modules from [`xdl`](https://github.com/expo/expo-cli/tree/main/packages/xdl/src).
- Does not implement [AppleDevice](https://github.com/expo/expo-cli/blob/main/packages/xdl/src/apple/AppleDevice.ts) or [CoreSimulator](https://github.com/expo/expo-cli/blob/main/packages/xdl/src/apple/CoreSimulator.ts) beta modules. Both were still behind flags and we can pull them in later as a way to reduce the initial review complexity.
- The majority of features are implemented using cross-platform logic as a way to reduce the amount of testing required to get full coverage (unclear if this is even achieved).
- Most code is doc blocked to add design clarity.
- Most 'iOS' stuff has been renamed to 'Apple' to account for the fact that Expo supports most platforms under the Apple umbrella.

<!--
How did you build this feature or fix this bug and why?
-->

# Test Plan

- Added a lot of unit testing.
- Heavy fixtures add to the PR line count.
- No E2E is currently implemented for simulator/emulator setup.

EvanBacon committed:
Added modules for interacting with Apple and Android platforms

EvanBacon committed:
Merge branch 'main' into @evanbacon/expo/cli/start/platform-code

EvanBacon commented on packages/expo/cli/start/platforms/android/__tests__/emulator-test.ts:
```diff
@@ -0,0 +1,82 @@
+  getAttachedDevicesAsync: jest.fn(),
+  isBootAnimationCompleteAsync: jest.fn(),
+
+  listDevicesAsync: jest.fn(() => Promise.resolve([])),
+  startDeviceAsync: jest.fn(() => Promise.resolve()),
```
```suggestion
  listDevicesAsync: jest.fn(async () => []),
  startDeviceAsync: jest.fn(async () => {}),
```

EvanBacon committed:
Update packages/expo/cli/start/platforms/android/__tests__/emulator-test.ts

EvanBacon committed:
Merge branch 'main' into @evanbacon/expo/cli/start/platform-code

EvanBacon committed:
Update CHANGELOG.md

EvanBacon committed:
Update packages/expo/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

ide commented on packages/expo/cli/start/platforms/AppIdResolver.ts:
```diff
@@ -0,0 +1,63 @@
+import { CommandError, UnimplementedError } from '../../utils/errors';
+import { get } from '../../utils/obj';
+
+export class AppIdResolver {
```
Add a docblock explaining this is for application identifiers like the Android application ID and Apple bundle identifier (as opposed to the EAS project ID)

EvanBacon committed:
Update AppIdResolver.ts

move mocks to the top

Update ApplePlatformManager.ts

ide commented on packages/expo/cli/start/platforms/ios/simctl.ts:
```diff
@@ -0,0 +1,281 @@
+/** Returns a list of devices which current state is 'Booted' as an array. */
+export async function getBootedSimulatorsAsync(): Promise<Device[]> {
+  const simulatorDeviceInfo = await getRuntimesAsync('devices');
+  return Object.values(simulatorDeviceInfo.devices).reduce((prev, runtime) => {
+    return prev.concat(runtime.filter((device) => device.state === 'Booted'));
+  }, []);
```
```suggestion
  return Object.values(simulatorDeviceInfo.devices).flatMap(
    (runtime) => runtime.filter((device) => device.state === 'Booted')
  );
```

ide commented on packages/expo/cli/start/platforms/ios/simctl.ts:
```diff
@@ -0,0 +1,281 @@
+  } catch (error: any) {
+    // Nov 15, 2020: Observed this can happen when opening the simulator and the simulator prompts the user to update the xcode command line tools.
+    // Unexpected token I in JSON at position 0
+    if (error.message.match('Unexpected token')) {
```
```suggestion
    if (error.message.includes('Unexpected token')) {
```

ide commented on packages/expo/cli/start/platforms/ios/simctl.ts:
```diff
@@ -0,0 +1,281 @@
+/** Return a list of iOS simulators. */
+export async function getDevicesAsync(): Promise<Device[]> {
+  const simulatorDeviceInfo = await getRuntimesAsync('devices');
+  return Object.values(simulatorDeviceInfo.devices).reduce(
+    (prev, runtime) => prev.concat(runtime),
+    []
+  );
```
```suggestion
  return Object.values(simulatorDeviceInfo.devices).flat();
```

ide reviewed (changes requested):
The mockClear calls can likely be replaced with one global clearMocks setting.

EvanBacon commented on packages/expo/cli/start/platforms/android/ADBServer.ts:
```diff
@@ -0,0 +1,105 @@
+  isRunning: boolean = false;
+
+  /** Returns the command line reference to ADB. */
+  getAdbExecutablePath(): string {
```
Good suggestion, I added this as a TODO so we can keep the PR minimal.

EvanBacon commented on packages/expo/cli/start/platforms/android/ADBServer.ts:
```diff
@@ -0,0 +1,105 @@
+
+  /** Execute an ADB command with given args. */
+  async runAsync(args: string[]): Promise<string> {
+    // await Binaries.addToPathAsync('adb');
```
Probably will have to move away from vendoring this since it would add a ton of bloat to each project.

EvanBacon commented on packages/expo/cli/start/platforms/android/AndroidPlatformManager.ts:
```diff
@@ -0,0 +1,52 @@
+      /** Expo Go URL. */
+      getExpoGoUrl: () => string | null;
+      /** Dev Client URL. */
+      getCustomRuntimeUrl: (props?: { scheme?: string }) => string | null;
```
We use "custom runtime" in a few places since this logic is generally agnostic from Expo and dev clients.

EvanBacon commented on packages/expo/cli/start/platforms/android/adb.ts:
```diff
@@ -0,0 +1,325 @@
+  return results;
+}
+
+/** Uninstall an app given its Android package name. */
```
adb requires the package name:
```
 uninstall [-k] PACKAGE
     remove this app package from the device
     '-k': keep the data and cache directories
```

EvanBacon committed:
PR Feedback

Drop manual mock clears

PR Feedback

Use spawnAsync instead of execSync

Update simctl.ts

Update ensureSimulatorAppRunning-test.ts

Use while loop

Added off hooks

Update activateWindow.ts

Update ensureSimulatorAppRunning.ts

EvanBacon committed:
lint fix

EvanBacon commented on packages/expo/cli/start/platforms/android/AndroidDeviceManager.ts:
```diff
@@ -0,0 +1,163 @@
+    } catch (error) {
+      let errorMessage = `Couldn't open Android app with activity "${launchActivity}" on device "${this.name}".`;
+      if (error instanceof CommandError && error.code === 'APP_NOT_INSTALLED') {
+        errorMessage += `\nThe app might not be installed, try installing it with: ${chalk.bold(
+          `npx expo run:android -d ${this.name}`
+        )}`;
```
We ensure Expo Go is installed automatically so we'd probably want to assert that before running the method. Currently we never use this method for opening on Expo Go.

ide commented on packages/expo/cli/utils/exit.ts:
```diff
@@ -1,7 +1,12 @@
     process.on(signal, listener);
   }
+  return () => {
+    for (const signal of killSignals) {
+      process.off(signal, listener);
+    }
```
Useful to let the listener be freed if the caller holds onto the function to remove the listeners even after removing them (matters more in website programming though)

```suggestion
    if (!listener) {
      return;
    }
    for (const signal of killSignals) {
      process.off(signal, listener);
    }
    // Allow the listener to be GC'd
    listener = null;
```

ide commented on packages/expo/cli/start/platforms/android/adbReverse.ts:
```diff
@@ -0,0 +1,60 @@
+
+export async function startAdbReverseAsync(ports: number[]): Promise<boolean> {
+  // Install cleanup automatically...
+  installExitHooks(() => {
```
Remove exit hooks when stopAdbReverseAsync is called

ide commented on packages/expo/cli/start/platforms/android/ADBServer.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+    // clean up
+    this.removeExitHook = installExitHooks(() => {
+      if (this.isRunning) {
+        this.stopAsync();
+      }
```
```suggestion
      if (this.isRunning) {
        this.stopAsync();
      }
      this.removeExitHook = () => {};
```

EvanBacon committed:
PR Feedback
